hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/lib/output/committer/manifest/ManifestCommitter.java

```diff
@@ -0,0 +1,712 @@
+   * @throws IOException never
+   */
+  @Override
+  public boolean isCommitJobRepeatable(final JobContext jobContext)
+      throws IOException {
+    return false;
+  }
+
+  /**
+   * Declare that task recovery is not supported.
+   * It would be, if someone added the code *and tests*.
+   * @param jobContext
+   *          Context of the job whose output is being written.
+   * @return false, always
+   * @throws IOException never
+   */
+  @Override
+  public boolean isRecoverySupported(final JobContext jobContext)
+      throws IOException {
+    return false;
+  }
+
+  /**
+   *
+   * @param taskContext Context of the task whose output is being recovered
+   * @throws IOException always
+   */
+  @Override
+  public void recoverTask(final TaskAttemptContext taskContext)
+      throws IOException {
+    LOG.warn("Rejecting recoverTask({}) call", taskContext.getTaskAttemptID());
+    throw new IOException("Cannot recover task "
+        + taskContext.getTaskAttemptID());
+  }
+
+  /**
+   * Commit the task.
+   * This is where the task attempt tree list takes place.
+   * @param context task context.
+   * @throws IOException IO Failure.
+   */
+  public void commitTask(final TaskAttemptContext context)
+      throws IOException {
+    ManifestCommitterConfig committerConfig = createCommitterConfig(true,
+        context);
+    StageConfig stageConfig = committerConfig.createJobStageConfig()
+        .withOperations(createStoreOperations())
+        .build();
+    taskAttemptCommittedManifest = new CommitTaskStage(stageConfig)
+        .apply("generate")
+        .getRight();
+    iostatistics.incrementCounter(COMMITTER_TASKS_COMPLETED, 1);
+    logCommitterStatisticsAtDebug();
+  }
+
+  /**
+   * Abort a task.
+   * @param context task context
+   * @throws IOException failure during the delete
+   */
+  public void abortTask(final TaskAttemptContext context)
+      throws IOException {
+    ManifestCommitterConfig committerConfig = createCommitterConfig(true,
+        context);
+    new AbortTaskStage(committerConfig.createJobStageConfig()
+        .withOperations(createStoreOperations())
+        .build())
+        .apply(false);
+    logCommitterStatisticsAtDebug();
+  }
+
+  /**
+   * This is the big job commit stage.
+   * Load the manifests, prepare the destination, rename
+   * the files then cleanup the job directory.
+   * @param jobContext Context of the job whose output is being written.
+   * @throws IOException failure.
+   */
+  @Override
+  public void commitJob(final JobContext jobContext) throws IOException {
+
+    ManifestCommitterConfig committerConfig = createCommitterConfig(false,
+        jobContext);
+
+    try (CloseableTaskSubmitter ioProcs =
+             committerConfig.createSubmitter()) {
+      // the stage config will be shared across all stages.
+      StageConfig stageConfig = committerConfig.createJobStageConfig()
+          .withOperations(createStoreOperations())
+          .withIOProcessors(ioProcs)
```

Comment:
Ioprocs are shared accross the whole CommitJobStage which includes LoadManifest and Rename. But in the docs it is written that, threadpool shouldn't be shared to avoid deadlocks.

Reply:
I should be more explicit. Each stage is executed in sequence within a task commit or job commit method; the thread pool is local to the commitTask/commitJob methods